### PR TITLE
[SPARK-23529][K8s] Support mounting hostPath volumes

### PR DESCRIPTION
**Pull request by madanadit**
## What changes were proposed in this pull request?

This PR introduces a new config `spark.kubernetes.driver/executor.volumes` taking a values of the format documented [here](https://docs.google.com/document/d/15-mk7UnOYNTXoF6EKaVlelWYc9DTrTXrYoodwDuAwY4/edit?usp=sharing)

The use case is to enable short-circuit writes to distributed storage on k8s. The Alluxio File System uses domain sockets to enable short-circuit writes from the client to worker memory when co-located on the same host machine. A directory, lets say /tmp/domain on the host, is mounted on the Alluxio worker container as well as the Alluxio client ( = Spark executor) container. The worker creates a domain socket /tmp/domain/d and if the client container mounts the same directory, it can write directly to the Alluxio worker w/o passing through network stack. The end result is faster data access when data is local.

## How was this patch tested?

Manual testing on a k8s v1.8 cluster. Unit tests added to `Driver/ExecutorPodFactorySuite`.

This PR replaces https://github.com/apache/spark/pull/21032.

### CODE REVIEW

**Comment by madanadit** [comment]
@liyinan926 @foxish Please take a look

**Comment by foxish** [comment]
jenkins, ok to test

**Comment by foxish** [comment]
@madanadit, thanks for following up with the PR and for your [design doc](https://docs.google.com/document/d/15-mk7UnOYNTXoF6EKaVlelWYc9DTrTXrYoodwDuAwY4/edit#heading=h.8jlem461uvwv). I had a few comments on the doc which I've added for discussion.

**Comment by madanadit** [comment]
Thanks @foxish for your feedback. As a first time contributor to Spark, I would like to limit the scope of the changes in this PR. Let me know when you're ready to review again.

**Comment by liyinan926** [comment]
My general comment is the code is centered around parsing and setting up `hostPath` volumes even though the proposal is generic. I would suggest making the code generic as well. Particularly I would like to see an implementation that is able to parse all volume related properties and add all discovered volumes in one shot.

**Comment by madanadit** [comment]
@liyinan926 Thanks for the comment. I refactored it to address your concern. The code is simple enough to be easily refactored if the need be by whoever implements the next volume type.

Note that the goal of this PR is limited to hostPath volumes. Even though the implementation here is general enough to be extended, I will not attempt to add other volume types in this PR.

**Comment by liyinan926** [comment]
Thanks @madanadit! I will take a look over the weekend or early net week.

**Comment by madanadit** [comment]
Hey @liyinan926, do you think you'll have time to take a look this week?

**Comment by foxish** [comment]
@madanadit, sorry about the long turnaround time here. I made a cursory pass over it - and it looks good. Before getting deeper into the code, I think we need more exhaustive unit testing for non-hostpath type volumes, and a couple of [e2e tests](https://github.com/apache-spark-on-k8s/spark-integration). Once those are done, I'm also happy to do a more detailed review here.

**Comment by madanadit** [comment]
Hi @foxish, I don't see why the 2 testing concerns should block reviewing this PR.
1. This PR does not attempt to address non-hostpath volumes (both implementation and unit tests are hence not included).
2. I agree e2e tests do make sense. However, e2e tests should not block this PR in any way. e2e tests can also be added once the PR is reviewed?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesVolumeUtils.scala
Comment (liyinan926) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesVolumeUtils.scala]: Looks like you don't really need this function as it's just a wrapper of `parseVolumesWithPrefix `.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala
Comment (liyinan926) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: Instead of putting the logic of volume mounting in `BasicDriverFeatureStep` and `BasicExecutorFeatureStep`, we should add a new step for mounting volumes, similarly to how we handle secrets, e.g., `MountVolumesFeatureStep` where the logic of `addVolumes` should be. This feature step can be used for both the driver and executors.

**Review by liyinan926 — CHANGES_REQUESTED** [review]

**Comment by andrusha** [comment]
@madanadit @liyinan926 @foxish I addressed comments here and added PVC support https://github.com/apache/spark/pull/21260. However I'm unsure if this is the right way to go, please check the PR.

**Comment by liyinan926** [comment]
@andrusha Is the purpose of https://github.com/apache/spark/pull/21260 to replace both this PR and https://github.com/apache/spark/pull/21238?

**Comment by andrusha** [comment]
@liyinan926 that's the idea. We should have a single step to mount all kinds of volumes including hostPath and emptyDir. The only problem is the configuration awkwardness.

**Comment by madanadit** [comment]
Thanks @liyinan926 for the review and @andrusha for addressing the comments and moving the work along. I can close this PR once #21260 merges.

**Comment by liyinan926** [comment]
@madanadit can you close this as #21260 has been merged?

**Pull request closed without merging**
